src/fsharp/vs/service.fs

```diff
@@ -2339,92 +2380,120 @@ type BackgroundCompiler(referenceResolver, projectCacheSize, keepAssemblyContent
                 Some (parseResults,checkResults)
             | _ -> None
 
-    /// Type-check the result obtained by parsing, but only if the antecedent type checking context is available. 
-    member bc.CheckFileInProjectIfReady(parseResults:FSharpParseFileResults,filename,fileVersion,source,options,textSnapshotInfo:obj option) =
-        reactor.EnqueueAndAwaitOpAsync("CheckFileInProjectIfReady " + filename, fun ct -> 
-          let answer = 
-            match incrementalBuildersCache.TryGetAny options with
-            | Some(Some builder, creationErrors, _) ->
-            
-                match bc.GetCachedCheckFileResult(builder,filename,source,options) with 
-                | Some (_parseResults, checkResults) -> Some (FSharpCheckFileAnswer.Succeeded checkResults)
-                | _ ->
-                match builder.GetCheckResultsBeforeFileInProjectIfReady filename with 
-                | Some tcPrior -> 
-        
+    /// 1. Asynchronously waits for file "lock"
+    ///
+    /// 2. Checks that cached results become available until we were waiting
+    ///
+    /// 3. Type checks the file
+    ///
+    /// 4. Records results in the cache
+    ///
+    /// 5. Starts whole project background compilation
+    member private bc.CheckOneFile
+        (parseResults: FSharpParseFileResults,
+         source: string,
+         fileName: string,
+         options: FSharpProjectOptions,
+         textSnapshotInfo : obj option,
+         fileVersion : int,
+         builder : IncrementalBuilder,
+         tcPrior : PartialCheckResults,
+         creationErrors : FSharpErrorInfo list) = 
+    
+        withFileCheckLock (fileName, options.ProjectFileName, fileVersion)
```

Comment:
This part does not look safe https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/vs/service.fs#L2334-L2338

What do you think?